napari/plugins/hookspecs.py

```diff
@@ -0,0 +1,51 @@
+"""
+All hook specifications for pluggable functionality should be defined here.
+
+A hook specification (hookspec) is a function signature (with documentation)
+that declares an API that plugin developers must adhere to when providing
+hook implementations (hookimpl).  Hook implementations provided by plugins (and
+internally by napari) will then be invoked in various places throughout the
+code base.
+
+hookspecs are a feature of pluggy:
+https://pluggy.readthedocs.io/en/latest/#specs
+
+These hookspecs also serve as the API reference for plugin developrs, so
+comprehensive documentation with complete type annotations is imperative.
+"""
+
+import pluggy
+from typing import Callable, Optional, List, Tuple, Union, Any, Dict
+
+hookspec = pluggy.HookspecMarker("napari")
+
+# layer data may be: (data,) (data, meta), or (data, meta, layer_type)
+# using "Any" for the data type for now
+LayerData = Union[Tuple[Any], Tuple[Any, Dict], Tuple[Any, Dict, str]]
+ReaderFunction = Callable[[str], List[LayerData]]
+
+
+@hookspec(firstresult=True)
+def napari_get_reader(path: str) -> Optional[ReaderFunction]:
+    """Return function capable of loading `path` into napari, or None.
+
+    This function will be called on File -> Open... or when a user drags and
+    drops a file/folder onto the viewer. This function must execute QUICKLY,
+    and should return ``None`` if the filepath is of an unrecognized format
+    for this reader plugin.  If the filepath is a recognized format, this
+    function should return a callable that accepts the same filepath, and
+    returns a list of layer_data tuples: Union[Tuple[Any], Tuple[Any, Dict]].
+
+    Parameters
+    ----------
+    path : str
+        Path to file, directory, or resource (like a URL)
+
+    Returns
+    -------
+    {function, None}
```

Comment:
this should be `function or None`

Reply:
oh yeah, those are def wrong :) thanks.
As a side note... I was looking around for thoughts on custom types (like `Optional[ReaderFunction]`) in numpy docstrings and stumbled on your previous conversation from https://github.com/napari/napari-core/issues/1.  Any updated thoughts on that?  Avoid still?

Reply:
@tlambert03 which part of it exactly?

Reply:
the question is whether it's still better to use the less descriptive:
```
    Returns
    -------
    Callable or None
```

instead of using PEP 484 type annotations defined higher up in the module, like:
```
    Returns
    -------
    ReaderFunction or None
```

Reply:
normally I would request it to be something like `None or callable of str -> tuple` but because it can be so complex, I think that using `ReaderFunction` is ok provided that it is defined in the docs and linked to from the signature

Reply:
because these are plugin developers, who should be consulting primarily online documentation, it should be fine, but I don't think it should ever be done on more intuitively discovered functions such as `add_*`, since then pulling up the docs in the interpreter would not be helpful

Reply:
@jni thoughts?